### PR TITLE
Scan locating and labelling

### PR DESCRIPTION
**Pull request by billetd**
# New Feature 
**Scan formation**

**module:** *scan.py*

**package:** *pydarn/utils*

## Scope
This PR adds rudimentary functionality to seperate a dmap_data list of libraries into scans, according to the scan flag specified within the records.

`build_scan` simply returns a 1D list with length equal to the number of dmap records, with an integer number indicating which scan each record is a part of. Initially this will be used for fan plots when merged, and so this functionality will be removed from PR #52. 

## Test code
```python
import pydarn

# Read in fitACF file. Works best with normalscan type data.
file = "20200507.0.fitacf"
SDarn_read = pydarn.SDarnRead(file)
fitacf_data = SDarn_read.read_fitacf()

# Sort into scans
beam_scan=pydarn.build_scan(fitacf_data)
print(beam_scan)
>>array([  1.,   1.,   1., ..., 720., 720., 720.])
```

### CODE REVIEW

File: pydarn/utils/scan.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+This module is is used for sorting a given dmap_data list of dictionaries
+into scans
+"""
+
+import numpy as np
+from typing import List
+
+def build_scan(dmap_data: List[dict]):
+    """
+    Returns list of size equal tonumber of records in dmap_data, with scan number
```
Comment (carleyjmartin) [inline pydarn/utils/scan.py:13]:
```suggestion
    Returns list of size equal to number of records in dmap_data, with scan number
```

File: pydarn/utils/scan.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    for each record
+
+    Parameters
+    ----------
+    dmap_data: List(dict)
+        list of records (dictionaries) representing dmap data
+    """ 
+    # Setup scans for easy locating
```
Comment (carleyjmartin) [inline pydarn/utils/scan.py:21]:
```suggestion
    # Set up scans for easy locating
```

File: pydarn/utils/scan.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    scan_mark = [sub['scan'] for sub in dmap_data]
+    no_scans = 0
+    beam_scan = np.zeros((len(dmap_data)))
+    for beam in range(len(dmap_data)):
+        if abs(scan_mark[beam]) == 1:
+            no_scans += 1
+            beam_scan[beam] = no_scans
+        if scan_mark[beam] == 0:
+            beam_scan[beam] = no_scans
+    no_scans += 1
```
Comment (carleyjmartin) [inline pydarn/utils/scan.py:32]: Is this line doing anything, you're already outside the for and returning beam_scan? Is it supposed to be inside the if?
```suggestion
            no_scans += 1
```

File: pydarn/utils/scan.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Copyright (C) 2020 SuperDARN Canada, Universtiy of Saskatchewan
+# Author(s): Daniel Billett
+"""
+This module is is used for sorting a given dmap_data list of dictionaries
```
Comment (carleyjmartin) [inline pydarn/utils/scan.py:4]:
```suggestion
This module is used for sorting a given dmap_data list of dictionaries
```

File: pydarn/utils/scan.py (hunk 3 above)
Comment (billetd) [reply pydarn/utils/scan.py:32]: Turns out it's actually redundant. Deleting.

**Comment by billetd** [comment]
Committed all the changes, so happy for this to be merged if no other issues found. This will actually work with other scan modes if the scan flags are set correctly in file, now I think about it.

A seperate PR will come after this is merged which will add the functionality to generate a "scan array", i.e. a n_beams x n_gates array with the chosen data in it (pwr, vel, etc...).

File: pydarn/utils/scan.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+def build_scan(dmap_data: List[dict]):
+    """
+    Returns list of size equal to number of records in dmap_data, with scan number
+    for each record
+
+    Parameters
+    ----------
+    dmap_data: List(dict)
+        list of records (dictionaries) representing dmap data
```
Comment (mts299) [inline pydarn/utils/scan.py:19]: add returns field in

File: pydarn/utils/scan.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+This module is used for sorting a given dmap_data list of dictionaries
+into scans
+"""
+
+import numpy as np
+from typing import List
+
+def build_scan(dmap_data: List[dict]):
+    """
+    Returns list of size equal to number of records in dmap_data, with scan number
```
Comment (mts299) [inline pydarn/utils/scan.py:13]: What is the purpose of this function? Sorry, I think I lost track of something.

File: pydarn/utils/scan.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    Parameters
+    ----------
+    dmap_data: List(dict)
+        list of records (dictionaries) representing dmap data
+    """ 
+    # Set up scans for easy locating
+    # Makes a list of size (number of records), with the scan number for each
+    scan_mark = [sub['scan'] for sub in dmap_data]
+    no_scans = 0
```
Comment (mts299) [inline pydarn/utils/scan.py:24]: Are no_scans supposed to be number of scans? 
maybe change to num_scans if so.

File: pydarn/utils/scan.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    beam_scan = np.zeros((len(dmap_data)))
+    for beam in range(len(dmap_data)):
+        if abs(scan_mark[beam]) == 1:
```
Comment (mts299) [inline pydarn/utils/scan.py:27]: Add a comment about why abs and why 1 vs. 0 so someone who is not familiar with scan flags can understand this.

**Comment by aburrell** [comment]
Just clocked this, and already had a similar function under development in a branch in pyDARNio.  Since this is a utility that is useful for more than plotting, perhaps we should have a discussion about where it should go.

**Comment by mts299** [comment]
@aburrell I think the functionality will be a bit different as @billetd is making this more specifically in mind with fan plots. Correct me if I am wrong, but I think your utility will be broader for generally scientific reasons. 

With that said, let us keep this functionality for now and merge the PR once all comments are addressed @billetd and let you make more progress on fan plots instead or waiting for pyDARNio to catch up. If, in the end, we put it in pyDARNio, then we have to modify the code assuming the same functionality. Plus, I don't think this will be expressed as a primary user utility in pyDARN, so that the deprecation will be more silent. 

People want fan plots bad, so we need to keep this a priority for pyDARN as we are not dedicated to pyDARNio now.

**Comment by aburrell** [comment]
> Correct me if I am wrong, but I think your utility will be broader for generally scientific reasons.

There are more functions, but there is also an essentially identical function.

> People want fan plots bad, so we need to keep this a priority for pyDARN as we are not dedicated to pyDARNio now.

Good enough for me!  Deprecation is easier than parallel development of dependencies.

**Comment by mts299** [comment]
@billetd *poke* *poke*

**Comment by billetd** [comment]
> @billetd _poke_ _poke_

Sorry, just got some revisions to do this week, then I'll get back on this.

File: pydarn/utils/scan.py (hunk 7 above)
Comment (billetd) [reply pydarn/utils/scan.py:29]: It's the current scan number used when iterating over all the records. I'll change to "current_scan"

**Comment by billetd** [comment]
All done with this now. As soon as it's merged, can go back to fan plotting branch.

**Pull request merged**